rfcs/widget-restructure-rfc-1.md

```diff
@@ -0,0 +1,554 @@
+    // ...
+  }
+  
+  pub fn add_widget<W: Widget>(&mut self, widget: W, widget_index: usize) -> &mut WidgetTreeBuilder {
+    // ...
+  }
+  
+  pub fn commit(self) {
+    // ...
+  }
+}
+```
+
+> Not super sold on the name of `WidgetTreeBuilder` since it's not really building anything?
+
+<br />
+
+<br />
+
+#### Constructed Widgets
+
+A constructed widget, for the purposes of this RFC, is one that is manually defined like above but uses some traits and derives to allow it to be used as a tagged (`</>`) component. Like the manually defined widgets, it's good for adding customization and extending the widget itself with other functionality.
+
+To integrate with the `rsx` tagging syntax without becoming a fully "functional widget," it must implement `WidgetConstructor` and define its props.
+
+```rust
+trait WidgetConstructor {
+  type Props;
+  fn create_props(base_props: BaseProps) -> Self::Props;
+  fn construct(id: Index, children: Children, props: Self::Props) -> Self;
+}
+
+/// A struct containing the base props most widgets expect.
+/// The widget can choose to use them or not.
+struct BaseProps {
+  styles: Option<Style>,
+  on_event: Option<OnEvent>
+}
+```
+
+> This trait can't be used as a trait object (`dyn WidgetConstructor`), but that's okay. It's only for ensuring the methods exist and are consistent.
+
+By having this bound on the widget struct, the `rsx` macro can then build the widget by doing something like:
+
+```rust
+let mut props = #widget_ident::create_props(base_props);
+#( props.#prop_ident = #prop_value );*
+let widget = #widget_ident::construct(widget_id, children, props);
+```
+
+Generated out might look something like:
+
+```rust
+// <MyWidget foo={Some(1)} bar={"Hello".to_string()} />
+let mut props = MyWidget::create_props(base_props);
+props.foo = Some(1);
+props.bar = "Hello".to_string();
+let widget = MyWidget::construct(widget_id, children, props);
+```
+
+As for the implementation of `Widget:;render`, it will look largely the same as the manual widgets. The main difference is that now props are no longer found directly on the struct.
+
+##### Derive
+
+One thing this allows us to *maybe* do is add a derive macro for automatically implementing `WidgetConstructor`.
+
+```rust
+#[derive(WidgetConstructor)]
+struct MyWidget {
+  #[widget_constructor(id)]
+  widget_id: Index,
+  #[widget_constructor(props = "MyProps")]
+  // Or `#[widget_constructor(props)]` to assume "<Widget Name>Props"
+  widget_props: MyProps,
+  #[widget_constructor(children)]
+  widget_children: Children,
+}
+
+#[derive(Default)]
+struct MyProps {
+  foo: Option<i32>,
+  bar: String
+}
+```
+
+A derive would be nice since in order to reduce boilerplate, but it's not necessary. And this may add an extra layer of confusion (due to required marker attributes and the `Default` bound on the props). But I left it here as a possible option and to see if there are other, better ways of implementing something like this.
```

Comment:
I like making life easier :D

Reply:
Yeah! I think this either needs to be cleaned up or just super well-documented for it to actually be easier haha